backend/tilavarauspalvelu/api/graphql/types/recurring_reservation/serializers/reschedule_serializer.py

```diff
@@ -0,0 +1,276 @@
+            raise ValidationError(msg, code=error_codes.RESERVATION_SERIES_NO_FUTURE_RESERVATIONS)
+
+        # We know there is at least one future reservation, so we can safely assume
+        # that we will have at least one reservation in the series
+        first_reservation: Reservation = self.instance.reservations.order_by("begin").first()  # type: ignore[attr-defined]
+        if begin_date != self.instance.begin_date and first_reservation.begin < now:
+            msg = (
+                "Reservation series' begin date cannot be changed after its "
+                "first reservation's start time is in the past."
+            )
+            raise ValidationError(msg, code=error_codes.RESERVATION_SERIES_ALREADY_STARTED)
+
+        RecurringReservation.validator.validate_series_time_slots(
+            begin_date=begin_date,
+            begin_time=begin_time,
+            end_date=end_date,
+            end_time=end_time,
+            reservation_start_interval=interval,
+        )
+        return data
+
+    def update(
+        self,
+        instance: RecurringReservation,
+        validated_data: ReservationSeriesRescheduleData,
+    ) -> RecurringReservation:
+        skip_dates: set[datetime.date] = set(self.initial_data.get("skip_dates", []))
+        buffer_time_before: datetime.timedelta | None = self.initial_data.get("buffer_time_before")
+        buffer_time_after: datetime.timedelta | None = self.initial_data.get("buffer_time_after")
+
+        had_access_codes = instance.reservations.requires_active_access_code().exists()
+
+        # Create both the recurring reservation and the reservations in a transaction.
+        # This way if we get, e.g., overlapping reservations, the whole operation is rolled back.
+        with transaction.atomic():
+            instance = super().update(instance, validated_data)
+            reservations = self.recreate_reservations(
+                instance,
+                buffer_time_before=buffer_time_before,
+                buffer_time_after=buffer_time_after,
+                skip_dates=skip_dates,
+            )
+
+        has_access_codes = instance.reservations.requires_active_access_code().exists()
+
+        # Must refresh the materialized view since reservations changed time.
+        if settings.UPDATE_AFFECTING_TIME_SPANS:
+            update_affecting_time_spans_task.delay()
+
+        if settings.SAVE_RESERVATION_STATISTICS:
+            create_or_update_reservation_statistics.delay(
+                reservation_pks=[reservation.pk for reservation in reservations],
+            )
+
+        EmailService.send_seasonal_reservation_modified_series_email(instance)
+
+        if had_access_codes or has_access_codes:
+            try:
+                PindoraService.reschedule_access_code(instance)
+            except PindoraNotFoundError:
+                PindoraService.create_access_code(instance, is_active=True)
```

Comment:
create upon `NotFound` only needed in the _ha**s**_ but not the _ha**d**_ scenario?